File: client/src/services/KeepsService.js
import { logger } from "@/utils/Logger.js"
import { api } from "./AxiosService.js"
import { Keep } from "@/models/Keep.js"
import { AppState } from "@/AppState.js"

class KeepsService{
  async deleteKeep(keepId) {
      const response = await api.delete(`api/keeps/${keepId}`)
      logger.log(response.data)
      const newKeeps = AppState.Keeps.filter(pojo => pojo['id'] !== keepId)
      logger.log(newKeeps)
      AppState.Keeps = newKeeps
  }

  async createKeep(keepData) {
    const response = await api.post("api/keeps", keepData)
    logger.log(response.data)
    AppState.Keeps.push(new Keep(response.data))
  }
  async getKeepById(keepId) {
    AppState.Keep = null
    const response = await api.get(`api/keeps/${keepId}`)
    logger.log(response.data)
    AppState.Keep = new Keep(response.data)
  }


  async getAllKeeps() {
    AppState.Keeps = []
    const response = await api.get("api/keeps")
    logger.log(response.data)
    const newKeep = response.data.map(keep => new Keep(keep)) ?? []
    AppState.Keeps = newKeep
  }

}

export const keepsService = new KeepsService()

File: server/Services/KeepsService.cs




using Microsoft.AspNetCore.Http.HttpResults;

namespace keeper.Services;

public class KeepsService
{
    private readonly KeepsRepository _keepRepo;

    public KeepsService(KeepsRepository keepRepo)
    {
        _keepRepo = keepRepo;
    }

    internal Keep CreateKeep(Keep keepData)
    {
        Keep newKeep = _keepRepo.CreateKeep(keepData);
        return newKeep;
    }

    internal string DeleteKeep(int keepId, Account userInfo)
    {
        Keep keep = GetKeepById(keepId);
        if (keep.CreatorId != userInfo.Id) throw new Exception("You cannot delete another user's keep");
        _keepRepo.DeleteKeep(keepId);
        return $"The keep with an id of '{keepId}' has been deleted!";
    }

    internal Keep EditKeep(Keep keepData, int keepId)
    {
        Keep originalKeep = GetKeepById(keepId);
        if (originalKeep.CreatorId != keepData.CreatorId) throw new Exception("You can not edit another user's keep");
        originalKeep.Name = keepData.Name ?? originalKeep.Name;
        originalKeep.Description = keepData.Description ?? originalKeep.Description;
        originalKeep.Img = keepData.Img ?? originalKeep.Img;

        _keepRepo.EditKeep(originalKeep);
        return originalKeep;
    }

    internal List<Keep> GetAllKeeps()
    {
        List<Keep> keeps = _keepRepo.GetAllKeeps();
        return keeps;
    }

    internal Keep GetKeepById(int keepId)
    {
        Keep keep = _keepRepo.GetKeepById(keepId);
        if (keep == null) throw new Exception($"Invalid id: {keepId}");
        return keep;
    }
}